xbmc/guilib/GUIRangesControl.cpp

```diff
@@ -0,0 +1,349 @@
+  m_guiUpperTexture(posX, posY, width, height, upperTexture),
+  m_iInfoCode(info)
+{
+  ControlType = GUICONTROL_RANGES;
+}
+
+CGUIRangesControl::~CGUIRangesControl() = default;
+
+void CGUIRangesControl::SetPosition(float posX, float posY)
+{
+  // everything is positioned based on the background image position
+  CGUIControl::SetPosition(posX, posY);
+  m_guiBackground.SetPosition(posX, posY);
+}
+
+void CGUIRangesControl::Process(unsigned int currentTime, CDirtyRegionList& dirtyregions)
+{
+  bool changed = false;
+
+  if (!IsDisabled())
+    changed |= UpdateLayout();
+
+  changed |= m_guiBackground.Process(currentTime);
+  changed |= m_guiOverlay.Process(currentTime);
+
+  for (auto& range : m_ranges)
+    changed |= range.Process(currentTime);
+
+  if (changed)
+    MarkDirtyRegion();
+
+  CGUIControl::Process(currentTime, dirtyregions);
+}
+
+void CGUIRangesControl::Render()
+{
+  if (!IsDisabled())
+  {
+    m_guiBackground.Render();
+
+    for (auto& range : m_ranges)
+      range.Render();
+
+    m_guiOverlay.Render();
+  }
+
+  CGUIControl::Render();
+}
+
+
+bool CGUIRangesControl::CanFocus() const
+{
+  return false;
+}
+
+
+void CGUIRangesControl::SetRanges(const std::vector<std::pair<float, float>>& ranges)
+{
+  ClearRanges();
+  for (const auto& range : ranges)
+    m_ranges.emplace_back(CGUIRange(m_guiLowerTexture, m_guiFillTexture, m_guiUpperTexture, range));
+}
+
+void CGUIRangesControl::ClearRanges()
+{
+  m_ranges.clear();
+  m_prevRanges.clear();
+}
+
+void CGUIRangesControl::FreeResources(bool immediately /* = false */)
+{
+  CGUIControl::FreeResources(immediately);
+
+  m_guiBackground.FreeResources(immediately);
+  m_guiOverlay.FreeResources(immediately);
+
+  for (auto& range : m_ranges)
+    range.FreeResources(immediately);
+}
+
+void CGUIRangesControl::DynamicResourceAlloc(bool bOnOff)
+{
+  CGUIControl::DynamicResourceAlloc(bOnOff);
+
+  m_guiBackground.DynamicResourceAlloc(bOnOff);
+  m_guiOverlay.DynamicResourceAlloc(bOnOff);
+
+  for (auto& range : m_ranges)
+    range.DynamicResourceAlloc(bOnOff);
+}
+
+void CGUIRangesControl::AllocResources()
+{
+  CGUIControl::AllocResources();
+
+  m_guiBackground.AllocResources();
+  m_guiOverlay.AllocResources();
+
+  m_guiBackground.SetHeight(25);
+  m_guiOverlay.SetHeight(20);
+
+  for (auto& range : m_ranges)
+    range.AllocResources();
+}
+
+void CGUIRangesControl::SetInvalid()
+{
+  CGUIControl::SetInvalid();
+
+  m_guiBackground.SetInvalid();
+  m_guiOverlay.SetInvalid();
+
+  for (auto& range : m_ranges)
+    range.SetInvalid();
+}
+
+bool CGUIRangesControl::UpdateColors()
+{
+  bool changed = CGUIControl::UpdateColors();
+
+  changed |= m_guiBackground.SetDiffuseColor(m_diffuseColor);
+  changed |= m_guiOverlay.SetDiffuseColor(m_diffuseColor);
+
+  for (auto& range : m_ranges)
+    changed |= range.SetDiffuseColor(m_diffuseColor);
+
+  return changed;
+}
+
+bool CGUIRangesControl::UpdateLayout()
+{
+  bool bChanged(false);
```

Comment:
Sometimes there's changed and sometimes bChanged, some consistency would be preferred

Reply:
good point